src/unwinder/find_fde/custom.rs

```diff
@@ -0,0 +1,135 @@
+use super::{FDEFinder, FDESearchResult};
+use crate::util::{deref_pointer, get_unlimited_slice};
+
+use core::sync::atomic::{AtomicBool, Ordering};
+use gimli::{BaseAddresses, EhFrame, EhFrameHdr, NativeEndian, UnwindSection};
+
+static mut CUSTOM_FDE_FINDER: Option<&(dyn FDEFinder + Sync)> = None;
+
+static CUSTOM_FDE_FINDER_INIT_STARTED: AtomicBool = AtomicBool::new(false);
+static CUSTOM_FDE_FINDER_INIT_COMPLETED: AtomicBool = AtomicBool::new(false);
+
+#[derive(Debug)]
+pub struct SetCustomFDEFinderError(());
+
+pub fn set_custom_fde_finder(
+    fde_finder: &'static (dyn FDEFinder + Sync),
+) -> Result<(), SetCustomFDEFinderError> {
+    if CUSTOM_FDE_FINDER_INIT_STARTED.swap(true, Ordering::SeqCst) {
+        Err(SetCustomFDEFinderError(()))
+    } else {
+        unsafe {
+            CUSTOM_FDE_FINDER = Some(fde_finder);
+        }
+        CUSTOM_FDE_FINDER_INIT_COMPLETED.store(true, Ordering::SeqCst);
+        Ok(())
+    }
+}
+
+fn get_custom_fde_finder() -> Option<&'static dyn FDEFinder> {
+    if CUSTOM_FDE_FINDER_INIT_COMPLETED.load(Ordering::SeqCst) {
+        Some(unsafe { CUSTOM_FDE_FINDER.unwrap() })
+    } else {
+        None
+    }
+}
+
+pub struct CustomFinder(());
+
+pub fn get_finder() -> &'static CustomFinder {
+    &CustomFinder(())
+}
+
+impl FDEFinder for CustomFinder {
+    fn find_fde(&self, pc: usize) -> Option<FDESearchResult> {
+        get_custom_fde_finder().and_then(|fde_finder| fde_finder.find_fde(pc))
+    }
+}
+
+pub trait BaseAddressFinder {
+    fn find_text_base(&self, _pc: usize) -> Option<usize>;
```

Comment:
How about
```rust
enum FrameInfoKind {
    EhFrameHdr(usize),
    EhFrame(usize),
}

struct FrameInfo {
    pub text_base: usize,
    pub kind: FrameInfoKind,
}
```

and a single method finding `FrameInfo` from PC? BTW this trait should be unsafe to implement.

Reply:
Good idea.

Done.